### PR TITLE
Show thumbnails in result lists

### LINKED ISSUE #1744: Add thumbnail to Search list view
**Is your feature request related to a problem? Please describe.**
In DSpace 7, thumbnails appear in Search Grid view but not List view. 
DSpace7_Search_grid
DSpace7_Search_list

**Describe the solution you'd like**
Please add the thumbnail to List view. This will create a view comparable to that in DSpace 6 XMLUI.

We like the current XMLUI search and browse lists because they combine thumbnail, title, author, year, search snippet or abstract snippet in a list that is easy to scroll through. We use then all the time to check and administer the repository, for example, using the thumbnail to confirm there is a file and that the thumbnail is correct, checking complete title, author, date, and indexing. It seems with the current DSpace 7 Search List and Grid Views, you can’t get all that in one view. 
DSpace6_XMLUI_Search_thumbnails
WorldCat.org is also a good model
WorldCat_search

**Describe alternatives or workarounds you've considered**

**Additional context**

The icon for list view suggests a thumbnail.
DSpace7_Search_view_icons

### LINKED ISSUE #1745: Add thumbnail to Browse
**Is your feature request related to a problem? Please describe.**
In DSpace 7 thumbnails do not appear in Browse view.
DSpace7_browse

**Describe the solution you'd like**
Please add the thumbnail to List view. This will create a view comparable to that in DSpace 6 XMLUI when webui.browse.thumbnail.show is configured in https://wiki.lyrasis.org/display/DSDOC6x/Configuration+Reference#ConfigurationReference-WEBUserInterfaceConfigurations.

We like the current XMLUI search and browse lists because they combine thumbnail, title, author, year, search snippet or abstract snippet in a list that is easy to scroll through. We use then all the time to check and administer the repository, for example, using the thumbnail to confirm there is a file and that the thumbnail is correct, checking complete title, author, date, and indexing. Users enjoy images.
DSpace6_XMLUI_browse

**Describe alternatives or workarounds you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by mspalti**
Fixes #1744 
Fixes #1745 

## Description
Introduces a new Angular configuration option to display Item thumbnail images in browse and search result lists.

## Instructions for Reviewers
If the new configuration option is set to be `true` you should see thumbnail images in browse and search results. (The default setting is false). To test, add the following to your config.dev.yml

`showItemThumbnails: true`

When true, the new configuration option tells the application to embed thumbnail info in the item reponse. This step was necessary for browse requests; thumbnail info is already included in search responses so there's no change in that case. 

The new configuration option also tells Angular when to show the thumbnail in the list view.

The effect is global. I looked at making a themeable version of `ItemSearchResultListElementComponent` since that might make it possible to add the thumbnail per individual collection. But I came to the (perhaps mistaken) conclusion that this component is not themeable.

List of changes in this PR:
* Added the new configuration option
* Initial list of "recent submissions" now requests the embedded thumbnail info regardless of configuration
* The browse components and browse service request embedded thumbnail only if so configured
* ItemSearchResultListElementComponent shows thumbnail  in item list if so configured

Sample:
Screen Shot 2022-08-21 at 12 08 30 PM

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [ ] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

File: config/config.example.yml
Comment (tdonohue) [inline config/config.example.yml:170]: I'd recommend we consider renaming this to align better with other configs.  Arguably, this setting only impacts two sections of the UI -- browse by & search.

So, I'd propose we consider moving this under the existing `browseBy` configs:
```
browseBy:
  ...
  showItemThumbnails: true
```

We could either document that it's also used by Search results... or we could implement a second config which is specific to search results (as I suspect we'll end up having more search configs anyways).
```
search:
  showItemThumbnails: true
```

In any case, my point is this isn't really a global configuration, so we should likely rename it to better describe the purpose.

I'd also recommend we enable it by default (as implied above) as I suspect most people will want thumbnails shown.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@mspalti : Thanks for this improvement!  Overall, the basics here work, but I did run into a few bugs and I have a few suggestions to improve the code.

First, the bugs...

1. When thumbnails are enabled, they mess up the alignment of some Entities in those same results list.  Here's an example from Browse By Title of the alignment when you have both Publications and Research Projects in the same list:
browse-by-title-alignment
I think there's two possible solutions here.  Either display "No Thumbnail Available" for all entity types, or just make sure there's an empty column placeholder so that the titles align properly.
2. Another bug I noticed is in the display of the "No Thumbnail Available" image in search results.  The text wraps very oddly. Also this screenshot shows an example of how Collections matching a search also get misaligned with Publications/Items in the same list.
search-alignment

As for the code itself, it looks good overall, but I did notice a few things that could use cleanup:
* There's an odd `src/app/collection-page.tar.gz` file added by this PR, likely by accident. It should be removed.
* I noticed there's no specs/tests for these changes.  Is there some way to at least provide some basic specs to verify the thumbnails appear when enabled & disappear when not enabled?  If it's not easy to do in basic specs, maybe we think about enhancing e2e tests for this PR?
* Finally, I have some suggestions on how to rename/refactor the configuration here (see inline below). 

Overall though, I'm very glad to see this work!  I don't think any of these requested changes should require massive code refactors (the largest request is likely the request for some specs).

File: config/config.example.yml
Comment (mspalti) [reply config/config.example.yml:170]: I'm looking into the "no thumbnails" text problem. This should ultimately be an easy fix but I'm trying to decide on the best approach. 

The problem is a little tricky because we have 3 different layouts for Item, search and browse views. Ideally text should just resize with the width of the thumbnail flex container but we may instead need to tap into the Angular component lifecycle and set the font-size after the layout is rendered...but I'd really prefer a pure css solution. (An svg image won't work because we do translations.)

Comment (mspalti) [reply config/config.example.yml:170]: In terms of config, to me it makes sense to have both search and browseBy configurations.

**Comment by mspalti** [comment]
@tdonohue , I think all problems you found are fixed. 

For the  "no thumbnail" font issue, I tried several things. I think the best approach is to add global classes for the image placeholders and then set the font-size based on the size of the parent container.  The sizes I've chosen seem to work ok for me.

For communities and collections in the search result list I'm adding an offset so the align correctly with other items.

For entities, I'm adding a thumbnail or using the default svg file when no thumbnail is available for the item.

Screen Shot 2022-09-09 at 1 53 44 PM

I also added some specs to be sure thumbnails are added/removed, etc.

**Comment by mspalti** [comment]
Oh, regarding config files I decided to add the thumbnail setting to the `browseBy` configuration and added a comment, per your first suggestion, @tdonohue . Two separate configuration locations for the behavior just seemed awkward to me on second look.

**Comment by mspalti** [comment]
OK. This should be ready for another review.

**Comment by mspalti** [comment]
I added thumbnail previews to mydspace items. Everything looks correct to me in what I've been able to preview thus far. A few features are things I haven't used yet and I wasn't able to preview them today (e.g. pooled tasks) so be on the lookout for issues there.

Also, components and tests were modified to inject the environment. So now this is indeed ready for review (assuming tests pass).

**Review by ybnd — CHANGES_REQUESTED** [review]
Overall everything works well, it's nice to have this be so easily configurable

* Looks like there's a stray file [here](https://github.com/DSpace/dspace-angular/pull/1780/files#diff-22de2ebb6553be68ad29235415b2474a1e0a308fd3a551c70d6dd1ed1ba84a34)
* I'm not really sure where detailed view is ever used, but this PR adds thumbnails to it as well and breaks the placeholders (I checked this manually via `&view=detailed`)
  
  


There's some more visual issues that become more apparent after this PR, but this is probably out of scope
* When first loading e.g. the search page in list view, thumbnails cause lots of content shifts while the load. I think we could solve this by surrunding the animation with a fixed frame, like the placeholders. I could tack this on to #1694 since that deals with the thumbnail component itself
* On `xs` screens, list view places the title & description below the thumbnail. This costs us a lot of information density; IMO we should mach the look on `s` screens here.

File: src/app/browse-by/browse-by-date-page/browse-by-date-page.component.ts
Comment (mspalti) [reply src/app/browse-by/browse-by-date-page/browse-by-date-page.component.ts:54]: Good suggestion. I've made this change.

File: config/config.example.yml
Comment (mspalti) [reply config/config.example.yml:172]: Agreed. I changed this to `showThumbnails`

File: src/app/core/browse/browse.service.ts
Comment (mspalti) [reply src/app/core/browse/browse.service.ts:32]: Good suggestion! I'll remember this for next time.

**Comment by mspalti** [comment]
@ybnd , about the detailed view, the code suggests that detailed views were designed only for MyDSpace search results. The `object-detail `components are all MyDspace previews.

There are a few layout issues with the MyDspace detailed view but that's out of scope for this PR since detailed previews always add `ds-thumbnail` when the object has an embedded thumbnail. This PR has no effect on that behavior.  Here's a screenshot of the MyDspace view.
Screen Shot 2022-09-13 at 12 49 00 PM

When you apply `view=detailed` to regular search lists the result is strange. The parent container `ds-object-detail` is used and the child components in the list are rendered in small columns. The layout is broken even without thumbnails. Here's a screenshot of search results without thumbnail images.
Screen Shot 2022-09-13 at 1 02 42 PM

It looks like the detailed layout was not intended to be used here...or maybe it's just unfinished. Either way I don't think this is something we need to worry about now. Feel free to correct me.

**Comment by mspalti** [comment]
OK. The mobile layout is now working as you suggested @ybnd . Thanks for that suggestion. I agree it's an improvement.

Screen Shot 2022-09-13 at 2 13 47 PM


I need to call out one additional change that I just made. It's bothered me that the "no thumbnail" placeholder max-width is `175px`. That seems ok in the Item view, even though it is larger than the thumbnails that dspace generates. But it's vexing in the list view because the thumbnails and svgs don't align well with the larger placeholder. I've set the `--ds-thumbnail-max-width` to what I think is a better size: `125px`.  

The thumbnail max-width has bothered me for some time ... but this change is only a suggestion. We could also set the max-width for result lists separately from `--ds-thumbnail-max-width`. 

Here's how things align now in the browse view, using the 125px max-width:
Screen Shot 2022-09-13 at 2 18 45 PM

**Comment by mspalti** [comment]
@ybnd , when I throttle to mid-tier mobile in Chome I can see the choppy rendering that you observed. Now that I see it I could work on it or apply a suggested fix. But it would also be great if you would want to tackle this with #1694 !

**Comment by ybnd** [comment]
Thanks for addressing everything!

If @tdonohue agrees I'll tackle the animation issue in #1694 (direct involvement because it makes changes in that component already & will make loading will take longer in some cases)

I did come accross one new problem though:
  * Thumbnails are also shown in relationship metadata representations, but are a bit broken
  * If we keep them there we should fix the font scaling and make sure they don't touch when the entries are short
  * For some reason the Entity placeholder SVGs aren't loaded and it falls back to the alt text

**Comment by ybnd** [comment]
The Item from that screenshot was from the demo site: https://demo7.dspace.org/entities/publication/5bb46321-21c6-44ce-b388-945500c87aba

**Comment by tdonohue** [comment]
@mspalti : Just a quick note...in case you hadn't seen it yet, this PR looks to be failing tests cause of lint issues in some of your `spec.ts` files.  It looks like you forgot a bunch of semi-colons at the end of lines.  You should be able to see these same lint errors locally by running `yarn lint`

**Comment by mspalti** [comment]
The layout for related entity thumbnails is working but the thumbnail is still missing from items.  So, the default svg and existing thumbnails aren't appearing yet. It looks to me like this is a backend issue. If we want a partial solution while we're looking into the backend, we could use the svg whenever the item thumbnail is missing and the `defaultImage` has been provided in `ds-thumbnail`. Here's the layout as it is now (with no default svg images):

Screen Shot 2022-09-14 at 6 57 02 PM

**Comment by ybnd** [comment]
Already a big improvement, but with the extra small thumbnails the placeholder text breaks down again on viewports under 1200px wide

I tried messing around a bit in `RelationshipService` & couldn't get thumbnail links to resolve from there (nesting `followLinks` doesn't seem to work in this case). We could add the thumbnails within `ds-related-items` [here](https://github.com/DSpace/dspace-angular/blob/6d361beb8812fa6c26e7c957ee2f4990e6612d6b/src/app/item-page/simple/related-items/related-items-component.ts#L65) as a workaround:
```
  getPage(page: number): Observable<RemoteData<PaginatedList<Item>>> {
    return this.relationshipService.getRelatedItemsByLabel(this.parentItem, this.relationType, Object.assign(this.options, { elementsPerPage: this.incrementBy, currentPage: page })).pipe(
      map((itemsRD: RemoteData<PaginatedList<Item>>) => {
        if (itemsRD.hasSucceeded && isNotEmpty(itemsRD.payload.page)) {
          itemsRD.payload.page.map(item => this.linkService.resolveLinks(item, followLink('thumbnail')));
        }
        return itemsRD;
      })
    );
  }
```
Checked locally on this branch & this resolves the missing SVG fallback images

**Comment by mspalti** [comment]
Thanks, @ybnd . I found the followLink signature this morning and then read your comment more carefully. Good solution. There are a few other methods in `relationship.service` that follow item links but they don't appear to require thumbnails.

Also, I'm glad to hear that it's a good idea to apply the `showThumbnail` configuration to services. It seemed like a good idea but I wasn't sure what others thought. I'll add the configuration here.

**Comment by tdonohue** [comment]
@mspalti : After merging #1791 (which is a larger, but important refactor), this now has some small merge conflicts.  I'd appreciate it if you could rebase this again when you get the chance.

**Comment by mspalti** [comment]
There was quite a lot of overlap with #1791 so I needed to merge instead of rebase and fix a few tests after the merge. Not sure why that was needed but all looks good now.

So this is ready for review again. 

I tested layouts using the device dimensions provided by Chrome tools and, so far as I can tell, everything looks good or at least acceptable. The text inside the thumbnail placeholder is always going to be tricky given the many permutations (viewport size, parent container size, translated text, etc.) but I mostly like what I'm seeing. I think I managed to look at all possible DSpace layouts, but do check me on that and let me know if we need more layout tweaks.

**Review by ybnd — CHANGES_REQUESTED** [review]

File: src/app/core/data/relationship-data.service.ts
Comment (mspalti) [reply src/app/core/data/relationship-data.service.ts]: Thanks for this @ybnd ! I sensed a code smell yesterday but had other distractions at work. I added `embedThumbnail` to `FindListOptions` (similar to how this is done for the `BrowseService)`. Now passing the parameter from` ds-related-items`.

Comment (ybnd) [reply src/app/core/data/relationship-data.service.ts]: Perfect!

If @tdonohue agrees I sense an opportunity to refactor some of the equivalent `SearchService` etc. methods to also handle thumbnail embeds in this way (instead of depending on custom `followLink`s each time).

IMO that makes sense as a separate "nice to have" PR, as everything here is functional _without_ it and this PR is already pretty extensive.

**Comment by mspalti** [comment]
I added configurable thumbnails for `ds-item-relationships` (that's one mentioned by @ybnd  that I hadn't seen before!) and verified that 'add relationship' modals are working. 

The only other code location I see that might be relevant here is `ds-dynamic-form-control-container`.  I don't know where to see this component in action but it looks like it's used with submissions. If it displays thumbnails then it needs the same approach used in `ds-item-relationships`. Just point me to where I can see the component in action if it needs work.

**Comment by mspalti** [comment]
I noticed that subject and author browses were failing because the thumbnail was being added to the request. This should be fixed.

**Comment by ybnd** [comment]
@mspalti About `ds-dynamic-form-control-container`: IIRC that would be for the list of related Items in the submission form itself, i.e. after you add relationships via the modal.

They're represented in a very compact way there, so we shouldn't add thumbnails IMO (or at least not until we can configure them separately from e.g. search/browse)

**Comment by mspalti** [comment]
Thanks for the screenshot. I agree that thumbnails aren't required.

**Comment by tdonohue** [comment]
@mspalti : Apologies, but this seems to have merge conflicts again.  If you get a chance, could you rebase it on `main`?  I think the minor conflicts were caused by merging #1771 (as this PR added a config to the same `browseBy` section in the configuration), but they should be easy to clean up.

**Comment by mspalti** [comment]
@tdonohue the merge conflict is fixed. Looks like the e2e failed for 16.x. I assume we just need to run the test again.

**Comment by tdonohue** [comment]
@mspalti : Yes, that looks like one of those semi-random e2e test failures. I just restarted the tests on Node 16.

**Comment by ybnd** [comment]
Looked everyhing over again, I'm +1 once we discuss/address one more issue:

Font scaling on Relationships on Item pages is still bugged for some "tablet" widths.

* Scaling the font down enough to fit would make the placeholder unreadable.
* Turning off thumbnails *only* here would require yet another Context & set of listable objects -- not great IMO.
* We could consider using SVG placeholders for Publication/Issue/Series/Journal entities (those are the only ones that could appear in this view in default DSpace)

**Review by tdonohue — APPROVED** [review]
👍 Thanks @mspalti !  The code looks good and my prior feedback has all been addressed.  

I tried to reproduce the font issue that @ybnd [reported above](https://github.com/DSpace/dspace-angular/pull/1780#issuecomment-1254803322).  I can *briefly* see it (it appears to exist in the server-side rendering of the page?) and then it disappears a moment later & I see the [same thing](https://github.com/DSpace/dspace-angular/pull/1780#issuecomment-1255383129) as @mspalti (where the font does scale down properly to really tiny, but readable).

It's most easy to reproduce by clicking reload (in your browser) on an OrgUnit page (or similar). When the page initially loads, the "No Thumbnail Available" text is overly large (like @ybnd 's screenshot), but a moment later it resizes (like @mspalti 's screenshot). It almost seems like the server-side rendering starts with larger text, and the resize occurs once things switch to client-side rendering.  At least this is the behavior for me on both Chrome & Firefox.   Because it resizes itself appropriately, I don't think this is something that should hold up this PR.... though maybe we can improve it in the future in a later PR.

So, all in all, I'm +1 getting this into 7.4.

**Comment by mspalti** [comment]
@tdonohue nice catch. I should have thought to test in production before calling it good! The element reference on the sever side doesn't have a width so the default (large) font was being chosen.

I added platform detection to the components. In the server context the font is hidden and appears correctly formatted on browser execution.

**Comment by ybnd** [comment]
@tdonohue @mspalti Now I'm not seeing the font issue either (and it wasn't only SSR last time - maybe I wasn't checking the latest version of the branch?)
+1

**Review by ybnd — APPROVED** [review]

**Review by tdonohue — APPROVED** [review]
👍 Gave this one last test again today & the SSR issue from yesterday is fixed!  So, this is all working perfectly now. Thanks again @mspalti for your hard work on this!

**Pull request merged**
